### DISCUSSION TITLE
Use sqlalchemy just as a SQL generator for different engines

### DISCUSSION THREAD

**Discussion by chamini2** [sqlalchemy/sqlalchemy#7459] [Usage Questions] [answered]
### Describe the use case

We are relying on a library which already has connections and authentication figured out and we just have to feed SQL statements to it based on some cases.

I wanted to see if sqlalchemy could help us generate said SQL, since it has already solved that part very well, but not try to open connections, etc.

For example, one use case I test for is a BigQuery connection:

```py
sqlalchemy.create_engine(f"bigquery://{db}")
```

The `db` part is important because the SQL generated uses that information.

But, this looks for the GOOGLE_APPLICATION_CREDENTIALS and fails if it does not find it:

```
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 576, in create_engine
    (cargs, cparams) = dialect.create_connect_args(u)
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/sqlalchemy_bigquery/base.py", line 806, in create_connect_args
    client = _helpers.create_bigquery_client(
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/sqlalchemy_bigquery/_helpers.py", line 52, in create_bigquery_client
    credentials, default_project = google.auth.default(scopes=SCOPES)
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/google/auth/_default.py", line 488, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
```

This makes sense in most cases, but for us, we really just want to sqlalchemy to generate SQL. Is there any way we can make this happen?

### Databases / Backends / Drivers targeted

all of them?

### Example Use

https://github.com/fal-ai/fal/blob/4469ca2a52ff6c8d679f6a9540de684aeecccf3e/src/faldbt/lib.py#L144-L150

### Additional context

For further details, we are using sqlalchemy engine to pass into a pandas dataframe SQLTable, to then get the SQL from it (kind of convoluted, I think, but it is what it is).

https://github.com/fal-ai/fal/blob/4469ca2a52ff6c8d679f6a9540de684aeecccf3e/src/faldbt/lib.py#L144-L150

**Comment by zzzeek** [comment, marked as answer]
well first that looks like a bug in the bigquery dialect.  you should be able to create an engine without it attempting to do any network calls, they shouldn't be doing any of that stuff until connect() is called, so first I would recommend reporting this to them as a bug.

from there, you'd need to dig into the "bigquery" package and find where they instantiate a Dialect() class, to work around this bug.    Then you can use the dialect object following the guidelines at https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#how-do-i-render-sql-expressions-as-strings-possibly-with-bound-parameters-inlined to render SQL strings using this dialect object.

**Reply by chamini2** [reply]
And as another workaround, can I just use create_mock_engine?

**Reply by chamini2** [reply]
The problem is, as you can see int he code block I linked:
```py
    engine = _alchemy_engine(adapter, target)
    pddb = pdsql.SQLDatabase(
        engine,
        meta=MetaData(engine, schema=target.schema),
    )
    pdtable = pdsql.SQLTable(target.name, pddb, data, index=False)
    alchemy_table: sqlalchemy.Table = pdtable.table.to_metadata(pdtable.pd_sql.meta)
```

this is for pandas, which specifically expects an engine.

**Reply by zzzeek** [reply]
possibly, im not sure if that codepath is called for mock_engine.

mock_engine.execute() is not going to work for DQL and DML statements though because there's no result set, it only works for DDL.  you can still use the dialect object though

**Reply by chamini2** [reply]
Using create_mock_engine does work! But indeed it is undesirable and have reported it https://github.com/googleapis/python-bigquery-sqlalchemy/issues/380 Thanks!
